options/options.go

```diff
@@ -737,17 +738,37 @@ func (opts *TerragruntOptions) DataDir() string {
 // AppendReadFile appends to the list of files read by a given unit.
 func (opts *TerragruntOptions) AppendReadFile(file, unit string) {
 	if opts.ReadFiles == nil {
-		opts.ReadFiles = map[string][]string{}
+		opts.ReadFiles = xsync.NewMapOf[string, []string]()
 	}
 
-	for _, u := range opts.ReadFiles[file] {
+	units, ok := opts.ReadFiles.Load(file)
+	if !ok {
+		opts.ReadFiles.Store(file, []string{unit})
+		return
+	}
+
+	for _, u := range units {
 		if u == unit {
 			return
 		}
 	}
 
 	opts.Logger.Debugf("Tracking that file %s was read by %s.", file, unit)
-	opts.ReadFiles[file] = append(opts.ReadFiles[file], unit)
+
+	// Atomic insert
+	// https://github.com/puzpuzpuz/xsync/issues/123#issuecomment-1963458519
+	_, _ = opts.ReadFiles.Compute(file, func(oldUnits []string, loaded bool) ([]string, bool) {
+		var newUnits []string
+
+		if loaded {
+			newUnits = append(make([]string, 0, len(oldUnits)+1), oldUnits...)
+			newUnits = append(newUnits, unit)
+		} else {
+			newUnits = []string{unit}
+		}
+
+		return newUnits, false
```

Comment:
```suggestion
		if !loaded {
			return []string{unit}, false
		}

		return append(oldUnits, unit), false
```

I thinks we can significantly simplify this. I assume it only makes sense to use `make` with a loop when we call `append` many times.

Reply:
From the explanation in the comment:

```
// It's important to allocate a new slice each time we mutate a 
// That's to avoid races with concurrent readers.
```

It's actually important to allocate a new slice each time instead of appending to the existing slice so that we don't experience a different race condition with concurrent reads.

Reply:
It's [here](https://github.com/puzpuzpuz/xsync/issues/123#issuecomment-1963458519).

I am not a go expert, so I might be mistaken as well, but I believe that `append` will only allocate a new slice if it has to because the capacity of the old slice is too small. If the old slice has enough capacity, it will just re-use the existing allocated slice and add another item.

Reply:
From the go docs:

```
// The append built-in function appends elements to the end of a slice. If
// it has sufficient capacity, the destination is resliced to accommodate the
// new elements. If it does not, a new underlying array will be allocated.
// Append returns the updated slice. It is therefore necessary to store the
// result of append, often in the variable holding the slice itself:
//
//	slice = append(slice, elem1, elem2)
//	slice = append(slice, anotherSlice...)
//
// As a special case, it is legal to append a string to a byte slice, like this:
//
//	slice = append([]byte("hello "), "world"...)
```

It only allocates new memory if the existing slice doesn't have enough capacity.

Reply:
Thanks for the godoc, I missed that condition. But anyway, in your case `append(oldUnits, unit)` will allocate new memory, if I understand correctly.

Reply:
`append(oldUnits, unit)` will only allocate new memory if there isn't enough capacity already.

`append(newUnits, unit)` is always appending to newly allocated memory because I explicitly allocate more with `make([]string, 0, len(oldUnits)+1)`